src/shell/autoexec.cpp

```diff
@@ -0,0 +1,675 @@
+		autoexec_blob_utf8.push_back(code_lf);
+	};
+
+	auto add_new_line_if_not_there = [&] {
+		// DOS line ending is CR+LF
+		if (autoexec_blob_utf8.back() != code_lf) {
+			add_new_line();
+		}
+	};
+
+	auto add_string = [&](const std::string& line) {
+		if (line.empty()) {
+			return;
+		}
+
+		bool should_skip_lf = false;
+		bool should_skip_cr = false;
+
+		for (const auto& character : line) {
+			if ((character == code_lf && should_skip_lf) ||
+			    (character == code_cr && should_skip_cr)) {
+				// This is a continuation of line ending mark
+				should_skip_lf = false;
+				should_skip_cr = false;
+				continue;
+			}
+
+			if (character == code_lf || character == code_cr) {
+				// This is a new line ending mark
+				should_skip_lf = (character == code_cr);
+				should_skip_cr = (character == code_lf);
+				add_new_line();
+				continue;
+			}
+
+			// This has to be an ordinary character
+			should_skip_lf = false;
+			should_skip_cr = false;
+			autoexec_blob_utf8.push_back(character);
+		}
+
+		// Enforce a DOS new line at the end if not present
+		add_new_line_if_not_there();
+	};
+
+	// Generate AUTOEXEC.BAT UTF-8 blob
+
+	autoexec_blob_utf8.clear();
+
+	bool prints_autogenerated  = false;
+	bool prints_config_section = false;
+	bool prints_variables      = false;
+	bool prints_first_line     = true;
+
+	static const std::string comment = ":: ";
+	static const std::string comment_generated =
+	        comment + MSG_Get("AUTOEXEC_BAT_AUTOGENERATED");
+	static const std::string comment_config_section =
+	        comment + MSG_Get("AUTOEXEC_BAT_CONFIG_SECTION");
+
+	if (should_insert_echo_off) {
+		add_string(comment_generated);
+		add_string("\n");
+		add_string("@ECHO OFF");
+		add_string("\n");
+
+		prints_first_line    = false;
+		prints_autogenerated = true;
+	}
+
+	for (const auto& [source, list] : autoexec_objects) {
+		if (list.empty()) {
+			continue;
+		}
+
+		switch (source) {
+		case Source::Variable:
+		case Source::BeforeConfigFile:
+		case Source::AfterConfigFile:
+			if (!prints_autogenerated) {
+				if (!prints_first_line) {
+					add_string("\n");
+				}
+				add_string(comment_generated);
+				add_string("\n");
+				prints_autogenerated  = true;
+				prints_config_section = false;
+			} else if (prints_variables && source != Source::Variable) {
+				// One-line separation between variables like
+				// 'BLASTER' or 'ULTRASND' and remaining
+				// autogenerated commands
+				add_string("\n");
+			}
+			break;
+		case Source::ConfigFile:
+			if (!prints_config_section) {
+				if (!prints_first_line) {
+					add_string("\n");
+				}
+				add_string(comment_config_section);
+				add_string("\n");
+				prints_autogenerated  = false;
+				prints_config_section = true;
+			}
+			break;
+		default: assert(false);
+		}
+
+		prints_variables  = (source == Source::Variable);
+		prints_first_line = false;
+
+		for (const auto& autoexec_object : list) {
+			add_string(autoexec_object->GetContent());
+		}
+	}
+}
```

Comment:
Ok, so this particular blob is a string 😄 Should be renamed to `create_utf8_autoexec_string` or something then. Also, I strongly recommend to not use globals. This function should be a pure function that simply returns a string, no?

Reply:
Function renamed, it is now visibly smaller. It cannot be fully pure - it has to access certain global data in some way - but I have changed it to return std::string.